src/combinatorics/set_family/slice.lean

```diff
@@ -0,0 +1,98 @@
+/-
+Copyright (c) 2021 Bhavik Mehta, Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta, Alena Gusakov, Yaël Dillies
+-/
+import data.fintype.basic
+
+/-!
+# `r`-sets and slice
+
+This file defines the `r`-th slice of a set family and provides a way to say that a set family is
+made of `r`-sets.
+
+An `r`-set is a finset of size `r`. The `r`-th slice of a set family is the set family made of its
+`r`-sets.
+
+## Main declarations
+
+* `finset.sized`: `A.sized r` means that `A` only contains `r`-sets.
+* `finset.slice`: `A.slice r` is the set of `r`-sets in `A`.
+
+## Notation
+
+`A # r` is notation for `A.slice r` in locale `finset_family`.
+-/
+
+open finset nat
+
+variables {α : Type*}
+
+namespace finset
+
+section sized
+variables {A B : finset (finset α)} {r : ℕ}
+
+/-! ### Sized -/
+
+/-- `sized r A` means that every finset in `A` has size `r`. -/
+def sized (r : ℕ) (A : finset (finset α)) : Prop := ∀ ⦃x⦄, x ∈ A → card x = r
+
+lemma sized.mono (h : A ⊆ B) (hB : B.sized r) : A.sized r := λ x hx, hB $ h hx
+
+lemma sized_union [decidable_eq α] : (A ∪ B).sized r ↔ A.sized r ∧ B.sized r :=
+⟨λ hA, ⟨hA.mono $ subset_union_left _ _, hA.mono $ subset_union_right _ _⟩,
+  λ hA x hx, (mem_union.1 hx).elim (λ h, hA.1 h) $ λ h, hA.2 h⟩
+
+alias sized_union ↔ _ sized.union
+
+variables [fintype α] {𝒜 : finset (finset α)} {s : finset α}
+
+lemma subset_powerset_len_univ_iff : 𝒜 ⊆ powerset_len r univ ↔ 𝒜.sized r :=
+forall_congr $ λ A, by rw mem_powerset_len_univ_iff
+
+alias subset_powerset_len_univ_iff  ↔ _ finset.sized.subset_powerset_len_univ
+
+lemma sized.card_le (h𝒜 : 𝒜.sized r) : card 𝒜 ≤ (fintype.card α).choose r :=
+begin
+  rw [fintype.card, ←card_powerset_len],
+  exact card_le_of_subset h𝒜.subset_powerset_len_univ,
+end
+
+end sized
+
+/-!
+### Slices
+
+The `r`-th slice of a set family the subset of its elements which have
+cardinality `r`.
+A few basic facts about slices.
+-/
+section slice
+variables {𝒜 : finset (finset α)} {A A₁ A₂ : finset α} {r r₁ r₂ : ℕ}
+
+/-- The `r`-th slice of a set family is the subset of its elements which have cardinality `r`. -/
+def slice (𝒜 : finset (finset α)) (r : ℕ) : finset (finset α) := 𝒜.filter (λ i, i.card = r)
+
+localized "infix ` # `:90 := finset.slice" in finset_family
+
+/-- `A` is in the `r`-th slice of `𝒜` iff it's in `𝒜` and has cardinality `r`. -/
+lemma mem_slice : A ∈ 𝒜 # r ↔ A ∈ 𝒜 ∧ A.card = r := mem_filter
+
+/-- The `r`-th slice of `𝒜` is a subset of `𝒜`. -/
+lemma slice_subset : 𝒜 # r ⊆ 𝒜 := filter_subset _ _
+
+/-- Everything in the `r`-th slice of `𝒜` has size `r`. -/
+lemma sized_slice : (𝒜 # r).sized r := λ _, and.right ∘ mem_slice.mp
+
+/-- Elements in distinct slices must be distinct. -/
+lemma ne_of_diff_slice (h₁ : A₁ ∈ 𝒜 # r₁) (h₂ : A₂ ∈ 𝒜 # r₂) : r₁ ≠ r₂ → A₁ ≠ A₂ :=
```

Comment:
Perhaps this should have both phrasings: `A₁ = A₂` (together with h1 and h2) implies `r₁ = r₂`, and then this version

Reply:
That's why I put the `pairwise_disjoint` phrasing. Is that enough to you?

Reply:
Using that would need a bunch of glue with `set.mem_univ` and unfolding of `disjoint`ness. Can we not have both?

Reply:
Sure, it's just not very canonical.